dash/include/dash/algorithm/Sort.h

```diff
@@ -0,0 +1,582 @@
+    PartitionBorder<T>& p_borders, std::vector<T>& partitions)
+{
+  DASH_ASSERT_EQ(
+      p_borders.is_stable.size(), partitions.size(),
+      "invalid number of partition borders");
+  // recalculate partition boundaries
+  for (std::size_t idx = 0; idx < partitions.size(); ++idx) {
+    // case A: partition is already stable -> done
+    if (p_borders.is_stable[idx]) continue;
+    // case B: we have the last iteration
+    //-> test upper bound directly
+    if (p_borders.is_last_iter[idx]) {
+      partitions[idx]          = p_borders.upper_bound[idx];
+      p_borders.is_stable[idx] = true;
+    }
+    else {
+      // case C: ordinary iteration
+      partitions[idx] =
+          (p_borders.lower_bound[idx] + p_borders.upper_bound[idx]) / 2;
+
+      if (partitions[idx] == p_borders.lower_bound[idx]) {
+        // if we cannot move the partition to the left
+        //-> last iteration
+        p_borders.is_last_iter[idx] = true;
+      }
+    }
+  }
+}
+
+template <typename ElementType, typename DifferenceType>
+inline std::pair<std::vector<DifferenceType>, std::vector<DifferenceType> >
+psort_local_histogram(
+    std::vector<ElementType> const& partitions,
+    ElementType const*              lbegin,
+    ElementType const*              lend)
+{
+  auto const nborders = partitions.size();
+  auto const sz       = partitions.size() + 2;
+  // Number of elements less than P
+  std::vector<DifferenceType> n_lt(sz, 0);
+  // Number of elements less than equals P
+  std::vector<DifferenceType> n_le(sz, 0);
+
+  auto const n_l_elem = std::distance(lbegin, lend);
+
+  std::size_t idx;
+
+  for (idx = 0; idx < nborders; ++idx) {
+    auto lb_it = std::lower_bound(lbegin, lend, partitions[idx]);
+    auto ub_it = std::upper_bound(lbegin, lend, partitions[idx]);
+
+    n_lt[idx + 1] = std::distance(lbegin, lb_it);
+    n_le[idx + 1] = std::distance(lbegin, ub_it);
+  }
+
+  n_lt[idx + 1] = n_l_elem;
+  n_le[idx + 1] = n_l_elem;
+
+  return std::make_pair(n_lt, n_le);
+}
+
+template <typename DifferenceType, typename ArrayType>
+inline void psort_global_histogram(
+    std::vector<DifferenceType> const& l_nlt,
+    std::vector<DifferenceType> const& l_nle,
+    ArrayType&                         g_nlt_nle)
+{
+  DASH_ASSERT_EQ(l_nlt.size(), l_nle.size(), "Sizes must match");
+  DASH_ASSERT_EQ(
+      g_nlt_nle.size(), (l_nlt.size() - 1) * 2, "Sizes must match");
+
+  /* reduce among all units */
+  std::fill(g_nlt_nle.lbegin(), g_nlt_nle.lend(), 0);
+
+  g_nlt_nle.team().barrier();
+
+  using glob_atomic_ref_t = dash::GlobRef<dash::Atomic<DifferenceType> >;
+
+  // TODO: Implement GlobAsyncRef<Atomic>, so we can asynchronously accumulate
+  for (std::size_t idx = 1; idx < l_nlt.size(); ++idx) {
+    // accumulate g_nlt
+    auto const g_idx_nlt      = (idx - 1) * 2;
+    auto       ref_clt        = g_nlt_nle[g_idx_nlt];
+    auto       atomic_ref_clt = glob_atomic_ref_t(ref_clt.dart_gptr());
+    atomic_ref_clt.add(l_nlt[idx]);
```

Comment:
This conversion through the DART gptr is really ugly (but of course necessary). Maybe we should come up with something more elegant, like `GlobRef<T>::atomic`.